spec/api-menu-item-spec.js

```diff
@@ -0,0 +1,402 @@
+          done()
+        }
+      }])
+      menu.delegate.executeCommand(menu, {}, menu.items[0].commandId)
+    })
+  })
+
+  describe('MenuItem with checked/radio property', () => {
+    it('clicking an checkbox item should flip the checked property', () => {
+      const menu = Menu.buildFromTemplate([{
+        label: 'text',
+        type: 'checkbox'
+      }])
+
+      assert.equal(menu.items[0].checked, false)
+      menu.delegate.executeCommand(menu, {}, menu.items[0].commandId)
+      assert.equal(menu.items[0].checked, true)
+    })
+
+    it('clicking an radio item should always make checked property true', () => {
+      const menu = Menu.buildFromTemplate([{
+        label: 'text',
+        type: 'radio'
+      }])
+
+      menu.delegate.executeCommand(menu, {}, menu.items[0].commandId)
+      assert.equal(menu.items[0].checked, true)
+      menu.delegate.executeCommand(menu, {}, menu.items[0].commandId)
+      assert.equal(menu.items[0].checked, true)
+    })
+
+    describe('MenuItem group properties', () => {
+      let template = []
+
+      beforeEach(() => {
+        for (let i = 0; i <= 10; i++) {
+          template.push({
+            label: `${i}`,
+            type: 'radio'
+          })
+        }
+
+        template.push({type: 'separator'})
+
+        for (let i = 12; i <= 20; i++) {
+          template.push({
+            label: `${i}`,
+            type: 'radio'
+          })
+        }
+      })
+
+      it('at least have one item checked in each group', () => {
+        const menu = Menu.buildFromTemplate(template)
+        menu.delegate.menuWillShow(menu)
+        assert.equal(menu.items[0].checked, true)
+        assert.equal(menu.items[12].checked, true)
+      })
+
+      it('should assign groupId automatically', () => {
+        const menu = Menu.buildFromTemplate(template)
+        const groupId = menu.items[0].groupId
+        for (let i = 0; i <= 10; i++) {
+          assert.equal(menu.items[i].groupId, groupId)
+        }
+        for (let i = 12; i <= 20; i++) {
+          assert.equal(menu.items[i].groupId, groupId + 1)
+        }
+      })
+
+      it("setting 'checked' should flip other items' 'checked' property", () => {
+        const menu = Menu.buildFromTemplate(template)
+
+        const checkValues = (start, end) => {
+          for (let i = start; i <= end; i++) {
+            assert.equal(menu.items[i].checked, false)
+          }
+        }
+
+        checkValues(0, 10)
+        menu.items[0].checked = true
+        assert.equal(menu.items[0].checked, true)
+
+        checkValues(1, 10)
+        menu.items[10].checked = true
+        assert.equal(menu.items[10].checked, true)
+
+        checkValues(0, 9)
+        checkValues(12, 20)
+        menu.items[12].checked = true
+        assert.equal(menu.items[10].checked, true)
+
+        checkValues(0, 9)
+        assert.equal(menu.items[12].checked, true)
+
+        checkValues(13, 20)
+      })
+    })
+  })
```

Comment:
The test ideas are great, but IMO the implementation is brittle. A lot of lines are spent dealing with hardcoded numbers, making it harder to change these tests in the future or to use as a starter pack for future tests that use a different template. These hardcoded assumptions also make the tests harder to read and to confirm they do what the description says.

Ideally the tests would minimize the implementation details, not hardcode assumptions, and express intent more clearly.

I'm sure this snippet could be improved, but here's a quick example of what I mean:


```js
// returns an array of { begin: x, end: y }
const findRadioGroups = (template) => {
  let groups = []
  let cur = null
  for (let i=0; i<=template.length; ++i) {
    if (cur && ((i===template.length) || (template[i].type!=='radio'))) {
      cur.end = i
      groups.push(cur)
      cur = null
    }
    else if (!cur && i<template.length && template[i].type==='radio'))) {
      cur = { begin: i }
    }
  }
  return groups
}

// returns array of checked menuitems in [begin,end)
const findChecked = (menuItems, begin, end) => {
  let checked = []
  for (let i=begin; i<end; ++i) {
    if (menuItems[i].checked) {
      checked.push(i)
    }
  }
  return checked
}


it('have exactly one item checked in each group', () => {
  const menu = Menu.buildFromTemplate(template)
  menu.delegate.menuWillShow(menu)

  const groups = findRadioGroups(template)
  for (let group of groups) {
    assert.equal(findChecked (menu.items, group.begin, group.end), [group.begin])
  }
})

it('should assign groupId automatically', () => {
  const menu = Menu.buildFromTemplate(template)

  let usedGroupIds = new Set()
  const groups = findRadioGroups(template)
  for (let group of groups) {
    const groupId = menu.items[group.begin]

    // groupId should be previously unused
    assert (!usedGroupIds.has(groupId))
    usedGroupIds.add(groupId)

    // everything in the group should have the same id
    for (let i=group.begin; i<group.end; ++i) {
      assert.equal(menu.items[i].groupId, groupId)
    }
  }
}

it("setting 'checked' should flip other items' 'checked' property", () => {
  const menu = Menu.buildFromTemplate(template)

  const groups = findRadioGroups(template)
  for (let group of groups) {
    assert.equal(findChecked(menu.items, group.begin, group.end), [])

    menu.items[group.begin].checked = true
    assert.equal(findChecked(menu.items, group.begin, group.end), [group.begin])

    menu.items[group.end-1].checked = true
    assert.equal(findChecked(menu.items, group.begin, group.end), [group.end-1])
  }
})
```